ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/RequestManager.java

```diff
@@ -16,43 +16,61 @@
 import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.messages.DisconnectMessage;
 
-import java.util.ArrayList;
+import java.math.BigInteger;
 import java.util.Collection;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Queue;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 
 public class RequestManager {
-  private final AtomicLong responseStreamId = new AtomicLong(0L);
-  private final Map<Long, ResponseStream> responseStreams = new ConcurrentHashMap<>();
+  private final AtomicLong requestIdCounter = new AtomicLong(0);
+  private final Map<BigInteger, ResponseStream> responseStreams = new ConcurrentHashMap<>();
   private final EthPeer peer;
+  private final boolean supportsRequestId;
 
   private final AtomicInteger outstandingRequests = new AtomicInteger(0);
 
-  public RequestManager(final EthPeer peer) {
+  public RequestManager(final EthPeer peer, final boolean supportsRequestId) {
     this.peer = peer;
+    this.supportsRequestId = supportsRequestId;
   }
 
   public int outstandingRequests() {
     return outstandingRequests.get();
   }
 
-  public ResponseStream dispatchRequest(final RequestSender sender) throws PeerNotConnected {
+  public ResponseStream dispatchRequest(final RequestSender sender, final MessageData messageData)
+      throws PeerNotConnected {
     outstandingRequests.incrementAndGet();
-    final ResponseStream stream = createStream();
-    sender.send();
+    final BigInteger requestId = BigInteger.valueOf(requestIdCounter.getAndIncrement());
+    final ResponseStream stream = createStream(requestId);
+    sender.send(supportsRequestId ? RequestId.wrapRequestId(requestId, messageData) : messageData);
     return stream;
   }
 
-  public void dispatchResponse(final EthMessage message) {
-    final Collection<ResponseStream> streams = new ArrayList<>(responseStreams.values());
+  public void dispatchResponse(final EthMessage ethMessage) {
+    final Collection<ResponseStream> streams = List.copyOf(responseStreams.values());
     final int count = outstandingRequests.decrementAndGet();
-
-    streams.forEach(s -> s.processMessage(message.getData()));
+    if (supportsRequestId) {
+      // If there's a requestId, find the specific stream it belongs to
+      final Map.Entry<BigInteger, MessageData> requestIdAndEthMessage =
+          RequestId.unwrapRequestId(ethMessage.getData());
+      Optional.ofNullable(responseStreams.get(requestIdAndEthMessage.getKey()))
+          .ifPresentOrElse(
+              responseStream -> responseStream.processMessage(requestIdAndEthMessage.getValue()),
+              // disconnect on incorrect requestIds
+              () -> peer.disconnect(DisconnectMessage.DisconnectReason.BREACH_OF_PROTOCOL));
```

Comment:
How does this happen in practice?  We have a response, but no open stream to receive the response, so we disconnect the target peer?  This seems likely to occur on our side rather than it being the fault of the remote ?

Reply:
We should have all the request ids for the outstanding requests we sent so if we receive a request id we're not tracking, that means the remote is breaching protocol somehow.

Reply:
isn't this our response to a request rather than a response to our request?  Seems to me if we have formulated a response but can't find the request id, that is a problem on our end and we shouldn't disconnect from the peer.  thoughts?

Reply:
This is handling their response to our request.

This is how we generate responses to remote requests:
https://github.com/RatanRSur/besu/blob/36d61b87c3a192d1b8bbd049230ba5a1ee00de08/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthProtocolManager.java#L278-L305